### PR TITLE
Coffee maker dev

### PR DESCRIPTION
**Pull request by tol8901**
Hello.

Check my new pull request, please.
In my branch coffee-maker-dev I've implemented tasks from the second sprint: 1 - 4, *5-*7 (from the shared table).

Best regards,
Anatolii Kravchenko.

### CODE REVIEW

File: application.rb [hunk 1]
```diff
@@ -0,0 +1,480 @@
+module Application
+  def self.run
+    begin
+      puts " Welcome to Coffee-Machine vendor menu ".center(50, '#')
+      # Manufacturers: DELONGHI, PHILIPS, REDMOND, SAECO.
+      manufacturers = ["DELONGHI", "PHILIPS", "REDMOND"]
+      puts "1) #{manufacturers[0]}"
```
Comment (dmitri-wm) [inline application.rb]: better to use each iteration `manufacturers.each  { ... } `

File: application.rb [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+      puts "2) #{manufacturers[1]}"
+      puts "3) #{manufacturers[2]}"
+      puts "4) Skip, leave devault (#{manufacturers[0]})"
+      print "Select a manufacturer, please: "
+      selected_vendor = gets.chomp.downcase
+
+      # Validation of user input
+      unless selected_vendor.gsub(/[1-4]/, '').empty?
```
Comment (dmitri-wm) [inline application.rb]: better not to use hardcoded values for checks  `/[1-4]/` , use dynamic condition

File: application.rb [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+        coffee_maker = Philips.new
+        break
+      when '3'
+        coffee_maker = Redmond.new
+        break
+      end
+    end while selected_vendor != '4'
+
+    # User selected vendor confirmation
+    if selected_vendor == '4'
```
Comment (dmitri-wm) [inline application.rb]: don't use magic numbers in condition change to dynamic OR use human readable value

File: application.rb [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+      sleep @loading_delay
+      puts "Goodbye :-)".rjust(50, ".")
+      puts "Press 'q' to quit"
+    end
+
+    def next_cup
+      @water_tank = 0 if @water_tank.nil?
+      @trash_can = 0 if @trash_can.nil?
+      @beans_can = 0 if @beans_can.nil?
+
```
Comment (dmitri-wm) [inline application.rb]: move initialization of variable into `initialize` method

File: application.rb [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+  end
+
+  class Cup
+    attr_accessor :volume, :is_full
+    def initialize
+      @is_full = false
+      @volume = 0
+    end
+
+    def cup_volume(volume=200)
```
Comment (dmitri-wm) [inline application.rb]: `def volume=()` already defined by `attr_accessor`

File: application.rb [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+      @volume = volume
+    end
+
+    def fill_cup(is_full=false)
```
Comment (dmitri-wm) [inline application.rb]: do we need default arg here?

File: application.rb [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+      when 'b' then coffee_maker.switch_off
+      when 'c' then coffee_maker.next_cup
+      end
+    end while choice != 'q'
+  end
+
+  class CoffeeMaker
+    def initialize
+      @is_turned_on = false # the state of the machine (On/Off)
+      @is_ready = false # the state of machine when it is ready to make a drink
```
Comment (dmitri-wm) [inline application.rb:83]: try to use state machine

File: application.rb [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+      # A can for fresh beans (0..1 kg)
+      @min_beans_can_amount = 0 # kilograms
+      @max_beans_can_amount = 1 # kilograms
+    end
+
+    def switch_on
+      @water_tank = 0 if @water_tank.nil?
+      @trash_can = 0 if @trash_can.nil?
+      @beans_can = 0 if @beans_can.nil?
+
```
Comment (dmitri-wm) [inline application.rb:92]: vars initialization in initializer

File: application.rb [hunk 9]
```diff
@@ -0,0 +1,480 @@
...
+      # The volume of the cup for drink, ml
+      @cup_volume = 200
+
```
Comment (dmitri-wm) [inline application.rb:96]: take water volume from passed`Cup` instance so volume of water will be decreased depending on `Cup` size

File: application.rb [hunk 10]
```diff
@@ -0,0 +1,480 @@
...
+      # Time of loading
+      @loading_delay = 3
```
Comment (dmitri-wm) [inline application.rb]: move to `initializer`

File: application.rb [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+
+      # Check state of the machine
+      if @is_turned_on # Turning on already working machine
```
Comment (dmitri-wm) [inline application.rb]: use StateMachine

File: application.rb [hunk 12]
```diff
@@ -0,0 +1,480 @@
...
+      @beans_can = 0 if @beans_can.nil?
+
+      # Check state of the machine (ON/OFF)
+      if @is_turned_on # The machine is turned on
+        user_select_drink
+      else # The machine is turned off
+        puts " First of all, Switch On the Machine ".center(50, "!")
+      end
+
+      self_check
```
Comment (dmitri-wm) [inline application.rb]: this check doesn't work?

File: application.rb [hunk 13]
```diff
@@ -0,0 +1,480 @@
...
+      # A can for fresh beans (0..1 kg)
+      @min_beans_can_amount = 0 # kilograms
+      @max_beans_can_amount = 1 # kilograms
+    end
+
+    def select_drink
+      begin
+        @chosen_drink = ""
+        puts "Available drinks:"
+        puts "a) Americano"
```
Comment (dmitri-wm) [inline application.rb]: 1) use initialized instance variable `@available_drinks` to keep hash like {'a' => 'Americano',  b' => "Espresso", ... }
2) use this variable to iterate and build menu
3) use keys for validation

File: application.rb (hunk 12 above)
Comment (tol8901) [reply application.rb]: Yes, it doesn't work.

Fixed this issue in the commit: 
[line: 139] Fix: Add self_check before making a drink

File: application.rb (hunk 1 above)
Comment (tol8901) [reply application.rb]: Fixed this issue in the commit:
[line: 7] Fix: Add iteration through the manufacturers list

File: application.rb (hunk 2 above)
Comment (tol8901) [reply application.rb]: Fixed this issue in the commit:
[line: 7] Fix: Add iteration through the manufacturers list

File: application.rb (hunk 3 above)
Comment (tol8901) [reply application.rb]: Changed magic number to the human readable value - 's' letter (first letter of the 'skip' word).

Fixed this issue in the commit:
[line: 7] Fix: Add iteration through the manufacturers list

File: application.rb (hunk 13 above)
Comment (tol8901) [reply application.rb]: Fixed this issue in the commit:
[line: 299] Fix: Implement iteration through a hash of available drinks

File: application.rb (hunk 6 above)
Comment (tol8901) [reply application.rb]: No, we don't need it.

File: application.rb (hunk 11 above)
Comment (tol8901) [reply application.rb]: Fixed this issue in the commit:
[line: 95] Fix: Implement State Machine

File: application.rb (hunk 7 above)
Comment (tol8901) [reply application.rb:83]: Fixed this issue in the commit:
[line: 95] Fix: Implement State Machine

File: application.rb (hunk 4 above)
Comment (tol8901) [reply application.rb]: Fixed this issue in the commit:
[line: 128] Fix: Changed main resources (Water, Trash, Beans) variables initialization

File: application.rb (hunk 8 above)
Comment (tol8901) [reply application.rb:92]: Fixed this issue in the commit:
[line: 128] Fix: Changed main resources (Water, Trash, Beans) variables initialization

File: application.rb (hunk 5 above)
Comment (tol8901) [reply application.rb]: Fixed this issue in the commit:
[line: 472] Fix: Remove redundant attr_accessor statement

File: application.rb (hunk 6 above)
Comment (tol8901) [reply application.rb]: Fixed this issue in the commit:
Implementation a Cup class and functionality

File: application.rb (hunk 9 above)
Comment (tol8901) [reply application.rb:96]: Fixed this issue in the commit:
Implementation a Cup class and functionality

File: application.rb (hunk 10 above)
Comment (tol8901) [reply application.rb]: Fixed this issue in the commit:
Implementation a Cup class and functionality

**Review by dmitri-wm — APPROVED** [review]

**Pull request merged**